File: program/Hoi4 Parser/tokenizer.cpp
#include "tokenizer.h"
#include "use_ex_log.h"
#include <format>
#include <fstream>

using namespace std;
using namespace hoi4::parser;

const Tokenizer::CompareChar Tokenizer::delimiter({ '\t', ' ', '\n', '\r', '#', '=', '>', '<', '}', '{', '"', '\0'});
const Tokenizer::CompareChar Tokenizer::blank({ '\t', ' ', '\n', '\r', '\0' });
const Tokenizer::CompareChar Tokenizer::endline({ '\n', '\r', '\0' });
const Tokenizer::CompareChar Tokenizer::marker({ '=', '>', '<', '}', '{' });
const char Tokenizer::note = '#';
const char Tokenizer::quote = '"';
const char Tokenizer::escape = '\\';

const char* fn_tkn = "tokenizer";

Tokenizer::Tokenizer(const char* filepath, token_list& tokens) :
    buffer(nullptr),
    buflen(0),
    bufpos(0),
    line(1),
    column(0),
    tree(nullptr),
    state(None),
    tokens(tokens)
{
    read_buf(filepath);
    tree = new ParseTree();
    while (bufpos <= buflen)
    {
        if (compose(buffer.get()[bufpos]))
        {
            auto _tree = tree->parse(elm);
            if (_tree == nullptr) // main-tree finish, go to next main-tree
            {
                cache_list();
                delete tree;
                tree = new ParseTree();
            }
            else { tree = _tree; }
        }
    }
    del_tree();
}

void Tokenizer::read_buf(const char* path)
{
    buflen = 0;
    bufpos = 0;
    ifstream fin;
    fin.open(path, ios::binary);
    if (!fin.is_open())
    {
        ex_log()->append(fn_tkn, format("could not open file: {}", path).c_str(), ExLog::ERR);
        buffer.reset(new char[1] {'\0'});
        return;
    }
    size_t starter = fin.get() == 0xEF && fin.get() == 0xBB && fin.get() == 0xBF ? 3 : 0; // remove BOM
    filebuf* fbuf = fin.rdbuf();
    buflen = (size_t)fbuf->pubseekoff(0, ios::end, ios::in) - starter;
    buffer.reset(new char[buflen + 1]);
    fbuf->pubseekpos(starter, ios::in);
    fbuf->sgetn(buffer.get(), buflen);
    fin.close();
    buffer.get()[buflen] = '\0';
}

void Tokenizer::cache_list()
{
    pToken _t = tree->once_get();
    if (_t == nullptr) { return; }
    tokens.push_back(ptok_u(_t));
}

bool Tokenizer::compose(char& ch)
{
    // while elm has been used its value will to be nullptr
                            // if elm didn't be used should pass it to next tree
    if (elm) { return true; }
    // some delimiters will bring to next loop 
                        // that won't use fget() to loop one more time to update state
    switch (state)
    {
    case Build_quo:
        if (ch == escape)
        {
            token << fget();
            state = Escape_quo;
            return false;
        }
        else if (ch == quote)
        {
            token << fget(); // keep the quote mark
            elm(token.str().c_str(), token.str().length(), line, column);   // will delete within process of tree->parse(...) when 
                                                            // not send to any Token,
                                                            // or will delete in Token::_vol_(...)
            state = None;
            return true;
        }
        else if (ch == endline)
        {
            token << quote;
            elm(token.str().c_str(), token.str().length(), line, column);
            state = None;
            return true;
        }
        token << fget();
        return false;
    case Escape_quo:
        if (ch == endline)
        {
            token << quote << quote;
            elm(token.str().c_str(), token.str().length(), line, column);
            state = None;
            return true;
        }
        else
        {
            token << fget();
            state = Build_quo;
            return false;
        }
    case Build_unquo:
        if (ch == delimiter)
        {
            elm(token.str().c_str(), token.str().length(), line, column);
            state = None;
            return true;
        }
        token << fget();
        return false;
    case Note:
        if (ch == endline) { state = None; }
        fget();
        return false;
    default:
        if (ch == quote)
        {
            token.clear();
            token.str("");
            token << fget(); // keep the quote mark
            state = Build_quo;
        }
        else if (ch == note)
        {
            state = Note;
            fget();
        }
        else if (ch == marker)
        {
            elm(fget(), line, column); // same as eToken above ^
            return true;
        }
        else if (ch == blank)
        {
            if (ch == '\n')
            {
                line++;
                column = 0;
            }
            else if (ch == '\t')
            {
                column += 3;
            }
            fget();
        }
        else
        {
            token.clear();
            token.str("");
            token << fget();
            state = Build_unquo;
        }
        return false;
    }
}

char Tokenizer::fget()
{
    char c = buffer.get()[bufpos];
    bufpos++;
    column++;
    return c;

}

void Tokenizer::del_tree()
{
    if (tree->get_from() != nullptr)
    {
        ex_log()->append(fn_tkn, format("interruption at line({}), column({})", line, column).c_str(), ExLog::WRN);
        tree->get_from()->append(tree->once_get());
        pTree _tree = tree->get_from();
        delete tree;
        tree = _tree;
        while (tree->get_from() != nullptr)
        {
            tree->get_from()->append(tree->once_get());
            _tree = tree->get_from();
            delete tree;
            tree = _tree;
        }
    }
    // since tree->once_get() only can get its build once
                // here try to cache one more time for interruption case
    cache_list();
    delete tree;
    tree = nullptr;
}

File: program/Parser/Tokenizer.cs
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Parser
{
    public class Tokenizer
    {
        static List<char> Delimiter = new() { '\t', ' ', '\n', '\r', '#', '=', '>', '<', '}', '{', '"', '\0' };
        static List<char> Blank = new() { '\t', ' ', '\n', '\r', '\0' };
        static List<char> EndLine = new() { '\n', '\r', '\0' };
        static List<char> Marker = new() { '=', '>', '<', '}', '{' };
        static char Note = '#';
        static char Quote = '"';
        static char Escape = '\\';

        enum States
        {
            None,
            Quotation,
            Escape,
            Unquotation,
            Note
        }
        States State { get; set; } = States.None;
        byte[] Buffer { get; set; }
        uint BufferPosition { get; set; } = 0;
        uint Line { get; set; } = 1;
        uint Column { get; set; } = 0;
        ParseTree Tree { get; set; } = null;
        Element Composed { get; set; } = null;
        StringBuilder Composing { get; set; } = new();
        List<TokenAPI> tokens { get; set; } = new();


        public Tokenizer(string FilePath, List<TokenAPI> Tokens)
        {
            ReadBuffer(FilePath);
            Tree = new ParseTree();
            while (BufferPosition < Buffer.Length)
            {
                if (Compose((char)Buffer[BufferPosition]))
                {
                    var _Tree = Tree.Parse(Composed);
                    if (_Tree == null)
                    {
                        CacheList();
                        Tree = new ParseTree();
                    }
                    else { Tree = _Tree; }
                }
            }
            EndCheck();
        }
        private void ReadBuffer(string FilePath)
        {
            using var file = new FileStream(FilePath, FileMode.Open, FileAccess.Read);
            if (file.ReadByte() == 0xEF && file.ReadByte() == 0xBB && file.ReadByte() == 0xBF)
            {
                file.Read(Buffer, 3, (int)file.Length - 3);
            }
            else
            {
                file.Read(Buffer);
            }
        }
        private bool Compose(char ch)
        {
            if (Composed.OwnValue == true) { return true; }
            switch(State)
            {
                case States.Quotation:
                    if (ch == Escape)
                    {
                        Composing.Append(GetChar());
                        State = States.Escape;
                        return false;
                    }
                    else if (ch == Quote)
                    {
                        Composing.Append(GetChar());
                        Composed = new(Composing.ToString(), Line, Column);
                        State = States.None;
                        return true;
                    }
                    else if (EndLine.Contains(ch))
                    {
                        Composing.Append(Quote);
                        Composed = new(Composing.ToString(), Line, Column);
                        State = States.None;
                        return true;
                    }
                    Composing.Append(GetChar());
                    return false;
                case States.Escape:
                    if (EndLine.Contains(ch))
                    {
                        Composing.Append(Quote);
                        Composing.Append(Quote);
                        Composed = new(Composing.ToString(), Line, Column);
                        State = States.None;
                        return true;
                    }
                    else
                    {
                        Composing.Append(GetChar());
                        State = States.Quotation;
                        return false;
                    }
                case States.Unquotation:
                    if (Delimiter.Contains(ch)) 
                    {
                        Composed = new(Composing.ToString(), Line, Column);
                        State = States.None;
                        return true;
                    }
                    Composing.Append(GetChar());
                    return false;
                case States.Note:
                    if (EndLine.Contains(ch))
                    {
                        State = States.None;
                    }
                    GetChar();
                    return false;
                default:
                    if (ch == Quote)
                    {
                        Composing.Clear();
                        Composing.Append(GetChar());
                        State = States.Quotation;
                    }
                    else if (ch == Note)
                    {
                        State = States.Note;
                        GetChar();
                    }
                    else if (Marker.Contains(ch))
                    {
                        Composed = new(GetChar().ToString(), Line, Column);
                        return true;
                    }
                    else if (Blank.Contains(ch))
                    {
                        if (ch == '\n')
                        {
                            Line++;
                            Column = 0;
                        }
                        else if (ch == '\t')
                        {
                            Column += 3;
                        }
                        GetChar();
                    }
                    else
                    {
                        Composing.Clear();
                        Composing.Append(GetChar());
                        State = States.Unquotation;
                    }
                    return false;
            }
        }
        private void CacheList()
        {
            var token = Tree.OnceGet();
            if (token == null) { return; }
            tokens.Add(token);
        }
        private char GetChar()
        {
            char ch = (char)Buffer[BufferPosition++];
            Column++;
            return ch;
        }
        private void EndCheck()
        {
            if (Tree.From != null)
            {
                Exceptions.Exception($"interruption at line({Line}), column({Column})");
                Tree.From.Append(Tree.OnceGet());
                Tree = Tree.From;
                while (Tree.From != null)
                {
                    Tree.From.Append(Tree.OnceGet());
                    Tree = Tree.From;
                }
            }
            CacheList();
        }
    }
}
